module/zfs/vfs_ratelimit.c

```diff
@@ -0,0 +1,664 @@
+{
+	int types, type;
+
+	if (rl == NULL) {
+		return (0);
+	}
+
+	types = 0;
+	for (type = ZFS_RATELIMIT_FIRST; type <= ZFS_RATELIMIT_LAST; type++) {
+		if (rl->rl_limits[type] > 0) {
+			types |= (1 << type);
+		}
+	}
+
+	return (types);
+}
+
+/*
+ * Returns the ratelimit structure that includes one of the requested types
+ * configured on the given dataset (os). If the given dataset doesn't have
+ * ratelimit structure for one of the types, we walk up dataset tree trying
+ * to find a dataset that has limits configured for one of the types we are
+ * interested in.
+ */
+static dsl_dir_t *
+ratelimit_first(objset_t *os, int types)
+{
+	dsl_dir_t *dd;
+	int mytypes;
+
+	ASSERT(RRM_READ_HELD(&os->os_spa->spa_ratelimit_lock));
+
+	dd = os->os_dsl_dataset->ds_dir->dd_ratelimit_root;
+	for (;;) {
+		if (dd == NULL) {
+			return (NULL);
+		}
+		mytypes = ratelimit_types(dd->dd_ratelimit);
+		if ((mytypes & types) != 0) {
+			/*
+			 * This dataset has at last one limit we are
+			 * interested in.
+			 */
+			return (dd);
+		}
+		if (dd->dd_parent == NULL) {
+			return (NULL);
+		}
+		dd = dd->dd_parent->dd_ratelimit_root;
+	}
+}
+
+/*
+ * Returns the ratelimit structure of the parent dataset. If the parent dataset
+ * has no ratelimit structure configured or the ratelimit structure doesn't
+ * include any of the types we are interested in, we walk up and continue our
+ * search.
+ */
+static dsl_dir_t *
+ratelimit_parent(dsl_dir_t *dd, int types)
+{
+	int mytypes;
+
+	ASSERT(RRM_READ_HELD(&dd->dd_pool->dp_spa->spa_ratelimit_lock));
+
+	for (;;) {
+		if (dd->dd_parent == NULL) {
+			return (NULL);
+		}
+		dd = dd->dd_parent->dd_ratelimit_root;
+		if (dd == NULL) {
+			return (NULL);
+		}
+		mytypes = ratelimit_types(dd->dd_ratelimit);
+		if ((mytypes & types) != 0) {
+			/*
+			 * This dataset has at last one limit we are
+			 * interested in.
+			 */
+			return (dd);
+		}
+	}
+}
+
+/*
+ * If we have any entries with 'timeslot > now' we also must have an entry with
+ * 'timeslot == now'. In other words if there is no entry with
+ * 'timeslot == now', it means that all the entires expired.
+ *
+ * We return either the most recent entry related to the given type or we return
+ * 'timeslot == now' entry not related to the given type and we will use it to
+ * store accouting information about this type as well.
+ */
+static struct rtslot *
+ratelimit_find(struct vfs_ratelimit *rl, int typebit, hrtime_t now)
+{
+	struct rtslot *slot;
+
+	ASSERT(MUTEX_HELD(&rl->rl_lock));
+
+	for (slot = list_head(&rl->rl_list); slot != NULL;
+	    slot = list_next(&rl->rl_list, slot)) {
+		if (slot->rts_timeslot < now) {
+			break;
+		}
+		if ((slot->rts_types & typebit) != 0 ||
+		    slot->rts_timeslot == now) {
+			return (slot);
+		}
+	}
+	/* All the entries expired. */
+#ifndef NDEBUG
+	for (slot = list_head(&rl->rl_list); slot != NULL;
+	    slot = list_next(&rl->rl_list, slot)) {
+		ASSERT(slot->rts_timeslot < now);
+	}
+#endif
```

Comment:
Should this remain in?

Reply:
Yes. If NDEBUG is not defined the ASSERT() is defined and we can then check that all the entries expired.